internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/wrpkit"
+)
+
+var (
+	ErrInvalidInput     = errors.New("invalid input")
+	ErrMisconfiguredQOS = errors.New("misconfigured QOS")
+)
+
+// Option is a functional option type for QOS.
+type Option interface {
+	apply(*Handler) error
+}
+
+type optionFunc func(*Handler) error
+
+func (f optionFunc) apply(c *Handler) error {
+	return f(c)
+}
+
+// Handler queues incoming messages and sends them to the next wrphandler
+type Handler struct {
+	next wrpkit.Handler
+	// queue for wrp messages
+	queue PriorityQueue
+
+	m sync.Mutex
+	// ingestQueue blocks if the queue ingestion is already in progress, i.e.: send
+	// as many queued messages as possible
+	ingestQueue chan bool
+	// shutdown shuts down the queue ingestion
+	shutdown context.CancelFunc
+}
+
+// New creates a new instance of the Handler struct.  The parameter next is the
+// handler that will be called and monitored for errors.
+func New(next websocket.Egress, opts ...Option) (*Handler, error) {
+	if next == nil {
+		return nil, ErrInvalidInput
+	}
+
+	h := &Handler{
+		next:        next,
+		ingestQueue: make(chan bool, 1),
+	}
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt.apply(h); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return h, nil
+}
+
+// Start starts the qos queue ingestion.
+func (h *Handler) Start() {
+	h.m.Lock()
+	if h.shutdown != nil {
+		return
+	}
+
+	// reset
+	var ctx context.Context
+	ctx, h.shutdown = context.WithCancel(context.Background())
+	h.m.Unlock()
+
+	go h.run(ctx)
+
+	select {
+	case h.ingestQueue <- true:
+		// send as many queued messages as possible
+	default:
+		// queue ingestion is already in progress
+	}
+
+}
+
+// Stop stops the qos queue ingestion.
+func (h *Handler) Stop() {
+	h.m.Lock()
+	shutdown := h.shutdown
+	// allows qos to restart
+	h.shutdown = nil
+	if shutdown == nil {
+		return
+	}
+	h.m.Unlock()
+
+	shutdown()
+}
+
+// HandleWrp is called to queue a message and then attempt to send as many queued messages as possible.
+func (h *Handler) HandleWrp(msg wrp.Message) error {
+	// queue newest message
+	h.queue.Enqueue(msg)
+	select {
+	case h.ingestQueue <- true:
+		// send as many queued messages as possible
+	default:
+		// queue ingestion is already in progress
+	}
+
+	return nil
+}
+
+// run sends as many queued messages as possible until the following occurs:
+// 1. the qos context was cancelled
+// 2. qos queue is empty
+// 3. there was a delivery failure
+// Note, Start() will automatically trigger a SendQueuedMessages().
+func (h *Handler) run(ctx context.Context) {
+	for {
+		select {
+		case <-h.ingestQueue:
+		// queue ingestion will start
+		case <-ctx.Done():
+			// QOS was stopped, exit
+			return
+		}
+
+		// always get the next highest priority message
+		// `ok` will stop the loop if the queue is empty
+		for msg, ok := h.queue.Dequeue(); ok; msg, ok = h.queue.Dequeue() {
+			select {
```

Comment:
I'd change the select to the following because it's simpler & doesn't have the bug the return creates of ending the go routune

```go
if ctx.Err() != nil {
  h.queue.Enqueue(msg)
  break
}
```